File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        return searchIndex(nums, target, 0, nums.size()-1);
    }
private:
    int searchIndex(vector<int>& nums, int& target, int l, int r) {
        int mid = (l + r) / 2;
        
        if (nums[l] <= nums[mid] && nums[mid] <= nums[r])
            return binarySearch(nums, target, l, r);
        else if (nums[l] <= nums[mid] && nums[mid] > nums[r])
        {
            if (nums[l] <= target && target <= nums[mid])
                return binarySearch(nums, target, l, mid);
            else return searchIndex(nums, target, mid+1, r);
        }
        else {
            if (nums[mid] <= target && target <= nums[r])
                return binarySearch(nums, target, mid, r);
            else return searchIndex(nums, target, l, mid-1);
        }

    }
    
    int binarySearch(vector<int>& nums, int& target, int l, int r) {
        int mid = (l + r) / 2;
        
        while (l <= r) {
            if (nums[mid] > target)  r = mid-1;
            else if (nums[mid] < target) l = mid+1;
            else return mid;
            mid = (l + r) / 2;
        }
        return -1;
    }
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def binary_search(self, nums, target):
        left, right = 0, len(nums)

        while left < right:
            pointer = (left + right) //2
            if nums[pointer] < target:
                left = pointer + 1
            elif nums[pointer] > target:
                right = pointer
            else:
                return pointer
        return -1


    def search(self, nums: List[int], target: int) -> int:
        max_idx = nums.index(max(nums))
        left, right = self.binary_search(nums[:max_idx+1], target), self.binary_search(nums[max_idx+1:], target)
        result = left if left != -1 else right + max_idx + 1 if right != -1 else -1
        return result
        